### PR TITLE
Tickets/dm 35399

### PR DESCRIPTION
**Pull request by fjammes**

### CODE REVIEW

File: argo-submit.sh [hunk 1]
```diff
@@ -55,4 +55,4 @@ if [ ! -d "$CFG_PATH" ]; then
 fi
 
 kubectl apply -k "$CFG_PATH"
-argo submit --serviceaccount=argo-workflow -p image="$INGEST_IMAGE" --entrypoint $entrypoint -vvv $DIR/manifests/workflow.yaml
+argo submit --serviceaccount=argo-workflow -p image="$INGEST_IMAGE" -p verbose=1 --entrypoint $entrypoint -vvv $DIR/manifests/workflow.yaml
```
Comment (andy-slac) [inline argo-submit.sh:58]: Is `verbose` still needed on master branch? (just checking that this is OK)

File: code-checks.sh [hunk 2]
```diff
@@ -41,12 +41,13 @@ if [ $# -ne 0 ] ; then
...
-  docker run -it -- "$INGEST_IMAGE" /ingest/bin/pytest.sh
+  # Build ingest image
+  $DIR/build.sh
+  docker run -- "$INGEST_IMAGE" /ingest/bin/pytest.sh
 fi
 if [ $mypy = true ]; then
-  docker run -it -- "$INGEST_IMAGE" mypy --config-file /ingest/python/mypy.ini /ingest/python/
+  mypy --config-file $DIR/rootfs/ingest/python/mypy.ini $DIR/rootfs/ingest/bin/replctl $DIR/rootfs/ingest/python/
+  # Command below require mypy to be installed in qserv-ingest
+  # docker run -it -- "$INGEST_IMAGE" mypy --config-file /ingest/python/mypy.ini /ingest/python/
```
Comment (andy-slac) [inline code-checks.sh]: Commented out stuff should probably be removed if line above works OK?

File: dev_local.sh [hunk 3]
```diff
@@ -35,15 +35,23 @@ if [ $# -ne 0 ] ; then
     exit 2
 fi
 
+echo -n "Check telepresence version "
+if ! telepresence version | grep "Client: v2"; then
```
Comment (andy-slac) [inline dev_local.sh]: Maybe `grep -q` to avoid any output?

File: env.in2p3-dp02.sh [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+# related directory must exists inside ./manifests/
+OVERLAY=in2p3-dp0.2
+
+if [ -n "$INGEST_RELEASE" ];
+then
+    TAG="$INGEST_RELEASE"
+else
+    GIT_HASH="$(git -C $DIR describe --dirty --always)"
+    TAG="$GIT_HASH"
+fi
```
Comment (andy-slac) [inline env.in2p3-dp02.sh]: This could be shortened to:

    TAG=${INGEST_RELEASE:-$(git -C $DIR describe --dirty --always)}

File: example/get_contrib_in_progress.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+Usage: `basename $0` [options] database
+
+  Available options:
+    -h          this message
+
+List contributions in 'IN_PROGRESS' state for a database
+EOD
+}
+
+kind=false
```
Comment (andy-slac) [inline example/get_contrib_in_progress.sh]: `kind` is not used anywhere below?

File: example/get_contrib_in_progress.sh [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+query="SELECT status,COUNT(*) FROM transaction_contrib WHERE \`database\`='$DATABASE' GROUP BY status;"
+kubectl exec -it  qserv-repl-db-0 -c repl-db -- \
+    mysql --host=qserv-repl-db-0 --user=qsreplica  qservReplica -e "$query"
+
+# Get detailed status for in-progress contributions
+query="SELECT id,transaction_id,worker,\`table\`,chunk,is_overlap,create_time,start_time,read_time,load_time
+  FROM transaction_contrib
+  WHERE \`database\`='$DATABASE' AND status='IN_PROGRESS';"
+kubectl exec -it  qserv-repl-db-0 -c repl-db -- \
+    mysql --host=qserv-repl-db-0 --user=qsreplica  qservReplica -e "$query"
```
Comment (andy-slac) [inline example/get_contrib_in_progress.sh]: newline is missing

File: example/get_contrib_read_failed.sh [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+Usage: `basename $0` [options] database
+
+  Available options:
+    -h          this message
+
+List 15 contributions in 'READ_FAILED' state for a database
+EOD
+}
+
+kind=false
```
Comment (andy-slac) [inline example/get_contrib_read_failed.sh]: This `kind` is also not used.

File: example/get_contrib_read_failed.sh [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+DATABASE="$1"
+LIMIT="LIMIT 15"
+
+# Get globals status for in-progress contributions
+query="SELECT id, url, transaction_id,worker,http_error,system_error,error \
+  FROM transaction_contrib \
+  WHERE \`database\`='$DATABASE' AND status='READ_FAILED' $LIMIT;"
+kubectl exec -it  qserv-repl-db-0 -c repl-db -- \
+    mysql --host=qserv-repl-db-0 --user=qsreplica  qservReplica -e "$query"
```
Comment (andy-slac) [inline example/get_contrib_read_failed.sh]: newline is missing

File: itest/install-qserv.sh [hunk 9]
```diff
@@ -6,7 +6,7 @@
 
 set -euxo pipefail
 
-OPERATOR_VERSION="2022.6.2-rc1-6-gef0de68"
+OPERATOR_VERSION="tickets/DM-35399"
```
Comment (andy-slac) [inline itest/install-qserv.sh]: Should this be changed to `main` before you merge (I think we should be removing branches after merge)?

File: itest/install-qserv.sh [hunk 10]
```diff
@@ -19,3 +19,6 @@ kubectl apply -f "$OPERATOR_DIR"/manifests/operator.yaml
 "$OPERATOR_DIR"/tests/tools/wait-operator-ready.sh
 kubectl apply -k "$OPERATOR_DIR"/manifests/base
 "$OPERATOR_DIR"/tests/tools/wait-qserv-ready.sh
+# Wait for replication controller initialization
+# see https://lsstc.slack.com/archives/G2JPZ3GC8/p1658864074354089
+sleep 20
```
Comment (andy-slac) [inline itest/install-qserv.sh:24]: 20 seconds looks arbitrary, is there a better way to check that service is running?

File: requirements-dev.txt [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+six==1.16.0
+snowballstemmer==2.2.0
+SQLAlchemy==1.4.39
+sqlalchemy-stubs==0.4
+tomli==2.0.1
+types-PyYAML==6.0.10
+types-requests==2.28.5
+types-urllib3==1.26.17
+typing_extensions==4.3.0
+urllib3==1.26.10
```
Comment (andy-slac) [inline requirements-dev.txt:33]: Is there a reason to pin all exact versions of dependencies? Same questions for `requirements.txt`

File: rootfs/ingest/bin/pytest.sh [hunk 12]
```diff
@@ -1,5 +1,8 @@
 #!/bin/sh
 
 # run unit tests inside qserv-ingest container
-pip install pytest
-pytest /ingest/python/ --capture=tee-sys -vv
+pip install pytest==7.1.2
```
Comment (andy-slac) [inline rootfs/ingest/bin/pytest.sh:7]: Also pinning a specific version? May be better to install latest one to avoid the need to update this file later?

File: rootfs/ingest/bin/replctl [hunk 13]
```diff
@@ -95,45 +98,56 @@ if __name__ == "__main__":
 
     args = parser.parse_args()
 
-    logger = util.get_default_logger(args.verbose)
+    env_verbose = os.getenv('QSERV_INGEST_VERBOSE')
+    if env_verbose is not None and len(env_verbose) != 0:
```
Comment (andy-slac) [inline rootfs/ingest/bin/replctl]: This can be shortened to:

    if env_verbose:

File: rootfs/ingest/python/qserv/contribution.py [hunk 14]
```diff
@@ -190,21 +195,26 @@ def monitor(self) -> bool:
...
+                    noretry_errmsg = "and is not retriable"
                 elif self.retry_attempts >= MAX_RETRY_ATTEMPTS:
-                    errmsg = "and has exceeded maximum number of ingest attempts"
-                if len(errmsg) != 0:
+                    noretry_errmsg = "and has exceeded maximum number of ingest attempts"
+                msg = (f"Contribution {self} is in status {contrib_monitor.status} "
+                       f"with error: {contrib_monitor.error}, "
+                       f"system error: {contrib_monitor.system_error}, "
+                       f"http error: {contrib_monitor.http_error}")
+                if len(noretry_errmsg) != 0:
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py]: In Python the canonical way to do it is:

    if noretry_errmsg:

File: rootfs/ingest/python/qserv/http.py [hunk 15]
```diff
@@ -82,9 +81,25 @@ def file_exists(url: str) -> bool:
...
-    """
-    Load json file at a given URL
+def json_get(base_url: str, filename: str) -> Dict:
+    """Load a JSON file located at a given URL
+
+    Parameters
+    ----------
+    base_url: `str`
+        JSON file location
+    filename `str`
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/http.py]: Needs colon after  `filename`

File: rootfs/ingest/python/qserv/http.py [hunk 16]
```diff
@@ -82,9 +81,25 @@ def file_exists(url: str) -> bool:
...
+        JSON file name
+
+    Raises
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/http.py:89]: `Raises` usually goes after `Returns`.

File: rootfs/ingest/python/qserv/http.py [hunk 17]
```diff
@@ -82,9 +81,25 @@ def file_exists(url: str) -> bool:
...
+    ------
+    IngestError:
+        Raise is URI scheme is not in http://, https://, file://
+
+    Returns
+    -------
+    Dict:
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/http.py]: `Dict` is a type, `Returns` should look like:

    data : `dict`
        JSON data ...

File: rootfs/ingest/python/qserv/http.py [hunk 18]
```diff
@@ -82,9 +81,25 @@ def file_exists(url: str) -> bool:
...
+        JSON data loaded inside a dictionnary
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/http.py]: dictionnary -> dictionary?

Maybe rephrase like "JSON data represented as a dictionary"?

File: rootfs/ingest/python/qserv/ingest.py [hunk 19]
```diff
@@ -91,52 +92,66 @@ def __init__(
...
         database = self.contrib_meta.database
         self.repl_client.database_publish(database)
 
-    def database_register_and_config(self, felis=None):
-        """
-        Register a database, its tables and its configuration inside replication system
-        using data_url/<database_name>.json as input data
+    def database_register_and_config(self,
+                                     replication_config: util.ReplicationConfig,
+                                     felis: Dict = None) -> None:
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/ingest.py]: Should be `Optional[Dict]` if default is `None`

File: rootfs/ingest/python/qserv/ingest.py [hunk 20]
```diff
@@ -91,52 +92,66 @@ def __init__(
...
+        """Register a database, its tables and its configuration inside replication/ingest system
+           using data_url/<database_name>.json as input data
+
+        Parameters
+        ----------
+        replication_config: `util.ReplicationConfig`
+            Configuration parameters for the database inside replication/ingest system
+        felis: `Dict, optional`
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/ingest.py]: `Dict` should be in backticks, but `optional` should not

File: rootfs/ingest/python/qserv/ingest.py [hunk 21]
```diff
@@ -186,27 +201,45 @@ def _build_contributions(
 
     def _ingest_transaction(self) -> bool:
         """Get contributions from a queue server for a given database
-        then ingest it inside Qserv,
-        during a super-transation
+        then ingest it inside Qserv during a super-transation
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/ingest.py]: transation -> transaction

File: rootfs/ingest/python/qserv/ingest.py [hunk 22]
```diff
@@ -186,27 +201,45 @@ def _build_contributions(
...
 
-        Returns:
-        --------
+        Raises
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/ingest.py]: `Raises` should appear after `Returns`

File: rootfs/ingest/python/qserv/jsonparser.py [hunk 23]
```diff
@@ -200,7 +190,28 @@ def get_regular_table_locations(responseJson: dict) -> List[Tuple[str, int]]:
...
+    """Retrieve database status inside JSON response from replication controller
+
+    Parameters
+    ----------
+    responseJson: `dict`
+        Response to a replication controller request, in json format
+    database: `str`
+        Database name
+    family: `str`
+        Family name of the database
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/jsonparser.py]: I think there should be a space before colon in parameters description.

File: rootfs/ingest/python/qserv/jsonparser.py [hunk 24]
```diff
@@ -200,7 +190,28 @@ def get_regular_table_locations(responseJson: dict) -> List[Tuple[str, int]]:
...
+
+    Raises
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/jsonparser.py]: Same thing, Raises goes after Returns.

File: rootfs/ingest/python/qserv/loadbalancerurl.py [hunk 25]
```diff
@@ -43,50 +43,75 @@
...
     """
 
-    count = 0
+    loadBalancerAlgorithm: Optional[LoadBalancerAlgorithm]
+    direct_url: str
 
-    def __init__(self, path: str, loadbalancers: List[str] = []):
-        """Manage a load balanced URL
+    def __init__(self, path: str, lbAlgo: LoadBalancerAlgorithm = None):
+        """Manage a load balanced URL for http:// protocole, also support access for file:// protocola
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/loadbalancerurl.py]: protocole, protocola -> protocol

File: rootfs/ingest/python/qserv/loadbalancerurl.py [hunk 26]
```diff
@@ -43,50 +43,75 @@
...
 
-        Args:
-            path (str): path of the url
-            loadbalancers (List[str], optional): List of http(s) load balancer urls. Defaults to [].
+        Parameters:
+        -----------
+            path str: path of the url
+            lbAlgo LoadBalancerAlgorithm: http(s) load balancer algorithm, not used for file:// access
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/loadbalancerurl.py]: Should look like:

    Parameters
    -----------
    path : `str`
        path of the url
    lbAlgo : `LoadBalancerAlgorithm`
        http(s) load balancer algorithm, not used for file:// access

File: rootfs/ingest/python/qserv/metadata.py [hunk 27]
```diff
@@ -118,14 +125,23 @@ def _filepath(self, filename: str) -> str:
...
 class TableSpec:
-    """Contain table specifications for a given database"""
+    """Contain table specifications
+
+    Parameters:
+    -----------
+        metadata_url: str
+            Url of metadata, used to access tables' json configuration files for R-I service
+        table_meta: Dict
+            metadata for a table
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/metadata.py]: De-indent by one level, space needed before colon.

File: rootfs/ingest/python/qserv/queue.py [hunk 28]
```diff
@@ -61,67 +65,103 @@ class QueueManager:
...
         else:
-            _LOG.warn("No table to load")
+            _LOG.warning("No table to load")
             self.current_table = None
 
-    def _select_locked_contribfiles(self) -> list[typing.Tuple[str, int, str, bool, str]]:
+    def all_succeed(self) -> bool:
+        """Check all contribution files have beed ingested successfully for current database
+
+        Returns:
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/queue.py]: Should be 

    Returns
    -------

File: rootfs/ingest/python/qserv/queue.py [hunk 29]
```diff
@@ -130,80 +170,136 @@ def insert_contribfiles(self):
...
+        """Assign contribfiles to a pod inside ingest queue
+
+        Parameters
+        ----------
+        contribfiles_to_lock_count: `int`
+            Maximum number of contribfile to lock
+
+        Returns
+        -------
+        contribfiles_locked_count: int
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/queue.py:248]: `int` needs backticks here too.

File: rootfs/ingest/python/qserv/queue.py [hunk 30]
```diff
@@ -212,47 +308,88 @@ def release_locked_contributions(self, ingest_success: bool) -> None:
...
+        """Retry failed update queries
+
+        Parameters
+        ----------
+            connection: `Any`
+                Sqlalchemy connection
+            query: `Any`:
+                Sql query
+            max_retry: int
+                Maximum num of retry attempts
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/queue.py]: De-indent by one level

File: rootfs/ingest/python/qserv/queue.py [hunk 31]
```diff
@@ -212,47 +308,88 @@ def release_locked_contributions(self, ingest_success: bool) -> None:
...
+                else:
+                    raise
+            except StatementError as ex:
+                util.check_raise(ex, sqlalchemy_retry_err_msg)
+                connection.rollback()
+                _LOG.error(
+                    "An error occurred during execution of a SQL statement: {},"
+                    " transaction has been rolled back (attempt #{} of {})".format(
+                        ex, retry_count, max_retry
+                    )
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/queue.py:393]: Logging should be using `%s` instead of `format`, same for above `_LOG.error`

File: rootfs/ingest/python/qserv/test_queue.py [hunk 32]
```diff
@@ -0,0 +1,360 @@
...
+    def create_schema(self) -> None:
+        self.db_meta.create_all(self.engine)
+
+    def empty_queue(self) -> None:
+        delete = self.queue.delete()
+        with self.engine.begin() as connection:
+            connection.execute(delete)
+
+    def count_contribfiles(self) -> int:
+        query = select([func.count("*")]).select_from(self.queue)
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/test_queue.py]: You can use `count()` for `COUNT(*)`.

File: rootfs/ingest/python/qserv/test_queue.py [hunk 33]
```diff
@@ -0,0 +1,360 @@
...
+        with self.engine.connect() as connection:
+            result = connection.execute(query)
+            contrib_count = next(result)[0]
+            result.close()
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/test_queue.py]: Use `result.scalar()`: https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=fetchone#sqlalchemy.engine.Result.scalar
(also in the methods below)

File: rootfs/ingest/python/qserv/test_queue.py [hunk 34]
```diff
@@ -0,0 +1,360 @@
...
+
+# ---------------------------------
+# Local non-exported definitions --
+# ---------------------------------
+
+_LOG = logging.getLogger(__name__)
+
+_CWD = os.path.dirname(os.path.abspath(__file__))
+
+_SCISQL_QUEUE_URL = 'sqlite:///testqservingest.db'
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/test_queue.py:59]: Do all test use the same database file? This probably means they cannot be run concurrently (e.g. with `pytest -n 10`)?

File: rootfs/ingest/python/qserv/test_replicationclient.py [hunk 35]
```diff
@@ -0,0 +1,75 @@
...
+    data_url = os.path.join(_CWD, "testdata", "dp01_dc2_catalogs")
+    contribution_metadata = metadata.ContributionMetadata(data_url)
+    tables_json_data = contribution_metadata.get_ordered_tables_json()
+
+    # Create empty credential file if not present
+    # TODO use http.AUTH_PATH instead of hard-coded value
+    auth_dir = os.path.join(Path.home(), ".lsst")
+    Path(auth_dir).mkdir(parents=True, exist_ok=True)
+    auth_file = os.path.join(auth_dir, "qserv")
+    Path(auth_file).touch()
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/test_replicationclient.py]: Modifying user $HOME directory in unit test looks odd. Can you use different location for http.AUTH_PATH?

File: rootfs/ingest/python/qserv/test_util.py [hunk 36]
```diff
@@ -0,0 +1,70 @@
...
+def test_ingestconfig() -> None:
+    """Check the function which compare files in two directories"""
+
+    parser = argparse.ArgumentParser(description="Test util module")
+    util.add_default_arguments(parser)
+
+    configfile = os.path.join(_CWD, "testdata", "dp02", "ingest.yaml")
+    args = parser.parse_args(['--config', configfile])
+
+    _LOG.debug(args.config.replication_config)
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/test_util.py]: Should probably be:

    _LOG.debug("%s", args.config.replication_config)

File: rootfs/ingest/python/qserv/util.py [hunk 37]
```diff
@@ -99,20 +98,75 @@ class IngestConfig:
...
+        as if it has never been configured for the database.
+        Default value: 120
+    """
+    cainfo: str = "/etc/pki/tls/certs/ca-bundle.crt"
+    ssl_verifypeer: int = 1
+    low_speed_limit: int = 60
+    low_speed_time: int = 120
+
+    def __post_init__(self) -> None:
+        """Set default value for all parameters, in case `None` value is used in constructor
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/util.py:156]: You could just define `__init__` and ignore `None` values.

File: rootfs/ingest/python/qserv/util.py [hunk 38]
```diff
@@ -180,7 +238,29 @@ def __call__(self, parser, namespace, values, option_string):
...
+def increase_wait_time(wait_sec: int) -> int:
     if wait_sec < 10:
         wait_sec *= 2
     return wait_sec
+
+
+def check_raise(e: Exception, not_raise_msgs: List[str]) -> None:
+    """Raise not recognized exceptions, depending on their message
+
+    Parameters:
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/util.py]: Should be

    Parameters
    ----------

**Review by andy-slac — APPROVED** [review]
Looks good, a bunch of minor comments.

File: argo-submit.sh (hunk 1 above)
Comment (fjammes) [reply argo-submit.sh:58]: Good catch! verbose=1 is info level. I should create a ticket to improve and document log management in the middle term.

File: code-checks.sh (hunk 2 above)
Comment (fjammes) [reply code-checks.sh]: :+1:

File: dev_local.sh (hunk 3 above)
Comment (fjammes) [reply dev_local.sh]: :+1:

File: env.in2p3-dp02.sh (hunk 4 above)
Comment (fjammes) [reply env.in2p3-dp02.sh]: :+1:

File: example/get_contrib_in_progress.sh (hunk 5 above)
Comment (fjammes) [reply example/get_contrib_in_progress.sh]: :+1:

File: rootfs/ingest/python/qserv/test_queue.py (hunk 34 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/test_queue.py:59]: Yes they cannot :+1: I added a comment in `pytest.sh`

File: rootfs/ingest/python/qserv/test_util.py (hunk 36 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/test_util.py]: :+1:

File: itest/install-qserv.sh (hunk 10 above)
Comment (fjammes) [reply itest/install-qserv.sh:24]: Hi need to wait for a @iagaponenko development to improve that, `wait-qserv-ready.sh` should do the job when the replication service will report to k8s its status.

File: rootfs/ingest/bin/pytest.sh (hunk 12 above)
Comment (fjammes) [reply rootfs/ingest/bin/pytest.sh:7]: I prefer pinning for now to avoid non-controlled updates/failures.

File: rootfs/ingest/python/qserv/contribution.py (hunk 14 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/contribution.py]: :+1:

File: requirements-dev.txt (hunk 11 above)
Comment (fjammes) [reply requirements-dev.txt:33]: I follow these good practices: https://12factor.net/dependencies, and prefer to manage dependencies version manually to ensure deterministic build.

File: rootfs/ingest/python/qserv/ingest.py (hunk 21 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/ingest.py]: :+1:

File: rootfs/ingest/python/qserv/ingest.py (hunk 19 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/ingest.py]: :+1:
